### PR TITLE
Performance improvements to FXMatrix and InterpolatedDoublesSurface

### PR DESCRIPTION
**Pull request by KirkWylie**

### CODE REVIEW

File: projects/OG-Analytics/src/main/java/com/opengamma/analytics/financial/forex/method/EmptyFXMatrix.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+  private static final long serialVersionUID = 1L;
+  private static final double[][] RATES_MATRIX = new double[0][0];
+  private static final Map<Currency, Integer> CURRENCIES_MAP = Collections.emptyMap();
+  
+  /**
+   * Singleton to avoid creating this unnecessarily.
+   */
+  public static final EmptyFXMatrix INSTANCE = new EmptyFXMatrix();
+  
+  public EmptyFXMatrix() {
```
Comment (jonathansenior) [inline projects/OG-Analytics/src/main/java/com/opengamma/analytics/financial/forex/method/EmptyFXMatrix.java]: Perhaps make this private to enforce use of the singleton instance, then also no need to override hashCode and equals.

File: projects/OG-Analytics/src/main/java/com/opengamma/analytics/financial/forex/method/ImmutableFXMatrix.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+package com.opengamma.analytics.financial.forex.method;
+
+import com.opengamma.util.money.Currency;
+
+/**
+ * A sub-class of {@link FXMatrix} that has strict immutabilty
+ * guarantees. This allows the {@link FXMatrix#copy()} operator
+ * to just return a handle to this instance.
+ */
+public class ImmutableFXMatrix extends FXMatrix {
```
Comment (jonathansenior) [inline projects/OG-Analytics/src/main/java/com/opengamma/analytics/financial/forex/method/ImmutableFXMatrix.java]: Also need to override `updateRates`

File: projects/OG-Analytics/src/main/java/com/opengamma/analytics/math/interpolation/GridInterpolator2D.java [hunk 3]
```diff
@@ -67,6 +70,39 @@ public GridInterpolator2D(final Interpolator1D xInterpolator, final Interpolator
...
+    
+    Map<Double, Interpolator1DDataBundle> result = new TreeMap<Double, Interpolator1DDataBundle>();
+    
+    ParallelArrayBinarySort.parallelBinarySort(xValues, yValues, zValues);
+    Double currX = xValues[0];
+    long currXBits = Double.doubleToRawLongBits(currX);
+    int currStartRange = 0;
+    for (int i = 0; i <= xValues.length; i++) {
+      if ((i == xValues.length) || (currXBits != Double.doubleToRawLongBits(xValues[i]))) {
+        //ParallelArrayBinarySort.parallelBinarySort(yValues, zValues, currStartRange, i - 1);
```
Comment (jonathansenior) [inline projects/OG-Analytics/src/main/java/com/opengamma/analytics/math/interpolation/GridInterpolator2D.java]: Commented out line?

Comment (KirkWylie) [inline projects/OG-Analytics/src/main/java/com/opengamma/analytics/math/interpolation/GridInterpolator2D.java]: It's WIP in that instance. I'll be able to resolve that and a few other things when the new interpolated surface is committed.

**Pull request merged**